### ISSUE TITLE
[dsicussion] Generic solutions for too-small-epsilon in FP16 training

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#35666]
Motivating PR: https://github.com/pytorch/pytorch/pull/33596

@albanD @ezyang Should we have a separate docs page (at least) about fp16 quirks, given that new built-in autocast support has been merged? Another related issue: https://github.com/pytorch/pytorch/pull/35594

It would also be nice to discuss some global hooks that can replace inf/nan in output/grad by zero, especially when a third-party library is used under the hood that doesn't allow to swap epsilons easily (or some epsilon factory, but that's more complicated).
Two related issues: https://github.com/pytorch/pytorch/issues/31829 https://github.com/pytorch/pytorch/issues/31557

**Comment by ezyang** [comment]
cc @mcarilli too

**Comment by gchanan** [comment]
On docs about fp16, that seems reasonable, particularly if @mcarilli helps writing them since he's an expert.

On global inf/NaN flag, we don't think it's that useful.  For grad, we already have the detect anomaly which is extremely slow so only useful for debugging.  But once you debug, you can write whatever value you want.  For the forward, we don't have an easy place to detect the anomalys like we do in backward.

For the "generic solution", we shouldn't be encoding exact epsilon values, but determining based on the dtype of the inputs.

**Comment by vadimkantorov** [comment]
> For the "generic solution", we shouldn't be encoding exact epsilon values, but determining based on the dtype of the inputs.

This would be nice!

Well, for things like https://github.com/pytorch/pytorch/pull/35594 some mitigation may be needed (maybe some python context to disable autocast locally).

> For the forward, we don't have an easy place to detect the anomalys like we do in backward.

Can't `register_forward_hook` or `register_pre_forward_hook` be used for that goal? If PyTorch provides some `torch.nn.utils.replace_by_zero(nan=True, pos_inf=False, neg_inf=False)` pre-made hook builder method, then a user has less place for a mistake.

A manual but standard `F.replace_by_zero(x, nan=True)` would also be nice. It could handle both forward and backward (or some variants may exist that do only grad or only forward).

> But once you debug, you can write whatever value you want.

I agree, but some standard builtin methods for this task would be nice, since the problem will be more common with fp16 autocasting in core. E.g. autocast docs now mention problems with not working `F.binary_cross_entropy` and suggesting to replace it with `F.binary_cross_entropy_with_logits`. It is possible for this case, but it would be nice to have a simple mitgation such as `F.replace_by_zero(F.binary_cross_entropy(...), grad_inf = 0)`. Even if a generic mitigation is not possible (no extension points or perf issues), some standard suggested mitigations is a great step forward.

E.g. LayerNorm was reported to have NaNs: https://discuss.pytorch.org/t/nan-in-layer-normalization/13846
It may be hard (and not sustainable) to debug and patch a library such as huggingface or fairseq to fix such things.

**Comment by albanD** [comment]
Another approach to this is to make kernels smarter (even though this might have perf implication:
```cpp
  // Example here would be clamping to a value >0 before taking the sqrt
  return AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(at::kHalf,
      result.scalar_type(), "eps", [result] {
            scalar_t eps = std::numeric_limits<at::scalar_value_type<scalar_t>::type>::min();
            result.clamp_min_(eps).sqrt_();
            return result;
      });
```

**Comment by vadimkantorov** [comment]
Will these templates work fine for half and bfloat16?

**Comment by albanD** [comment]
Yes, we define numeric_limits info for our custom types.
We might even be able to add a default eps value in there?
Then wrapping this into a nice function could allow you to get the default eps to use for a given type easily.

**Comment by vadimkantorov** [comment]
This may work for many cases but not for all cases - but having a standard idiom for this in the codebase would be nice. Different operations may require different eps, e.g. some operations may underflow when called with a legitimate eps for the type (e.g. `log(x)`?)

Suggestions about standard functions F.replace / wrapper modules / functools wrapper - can be helpful suggested mitigations for the nan gradient case while it's not fixed in third-party / core ops. E.g. torch.where probably still returns nan gradient occasionally

**Comment by vadimkantorov** [comment]
clamp_min is smart because it produces zero gradient in those areas automatically!

**Comment by albanD** [comment]
If you autodiff, any replace function will also do that. Because if the original value is not used, the gradient is 0.

> Different operations may require different eps, e.g. some operations may underflow when called with a legitimate eps for the type (e.g. log(x)?)

There shouldn't be too many of these right? Hopefully we should be able to define all of these?
